### PR TITLE
Kayla Ecker - Carets

### PR DESCRIPTION
**Pull request by kaylaecker**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |   Deciding how to separate the classes and to put them all in one file rather than separate ones. I figured if it ended up being too confusing I could split them up, but it ended up working out fine and I did not need to take extra steps in calling to additional rb files.    |
| Describe a concept that you gained more clarity on as you worked on this assignment. |  Definitely on the design and what I would need to put where (what information under which classes). I ended up reorganizing quite a bit so that things would make sense and not be in a confusing order.  |
| Describe a _nominal test_ that you wrote for this assignment. | checking initialize responses   |
| Describe an _edge case test_ that you wrote for this assignment. | checking when dates overlapped vs start/end date being the same .. apparently edge cases don't give any additional coverage percentage.    |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? | It was a big help in organizing code and getting myself to split things up into more manageable pieces.   |

### CODE REVIEW

File: lib/hotel.rb [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+    attr_reader :start_date, :end_date
+
+    def initialize(start_date, end_date)
+
+      @start_date = start_date
+      @end_date = end_date #should this be under the if as an else?
+
+      if @start_date > @end_date
+        raise ArgumentError, "Start date cannot come after end date"
```
Comment (CheezItMan) [inline lib/hotel.rb:34]: Shouldn't this be `raise ArgumentError.new("Start date cannot come after end date")`

File: specs/hotel_spec.rb [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+
+      @date.must_respond_to :end_date
+      @date.end_date.must_be_kind_of Date #edge1
+      @date.end_date.must_equal @end_date #edge2
+    end
+
+    it "checks initialize when start date is after end date" do
+      start_date = Date.new(2018,1,1)
+      end_date = Date.new(2018,1,3)
+      proc {DateRange.new(end_date, start_date)}.must_raise Exception
```
Comment (CheezItMan) [inline specs/hotel_spec.rb:29]: You should be checking for specific errors being raised like `ArgumentError`.

File: lib/hotel.rb [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+      @room_reservation[room_number] << new_reservation
+      if @date_reserve.key?(date.start_date)
+        @date_reserve[date.start_date] << new_reservation
+      else
+        @date_reserve[date.start_date] = [new_reservation]
+      end
+
+    end
+
+      def get_all_rooms
```
Comment (CheezItMan) [inline lib/hotel.rb:181]: Shouldn't this just be an accessor with `attr_reader`

File: lib/hotel.rb [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+      end
+
+      def get_open_rooms(date)
+        open_rooms = []
+        20.times do |i|
+          open_rooms[i] = true
+        end
+
+        current_date = date.start_date
+        while current_date < date.end_date do
```
Comment (CheezItMan) [inline lib/hotel.rb:192]: This type of comparison would make an excellent method in your `DateRange` class.  It's something you do multiple times so a method like this is very appropriate.

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| More commits would be better.  Also commit messages that describe what was accomplished and not oops I...    |
Answer comprehension questions	| Check, I hope you figured out that splitting things into separate files makes it easier.  Also good observation about edge cases and test coverage. |
**Design** |  | 
Each class is responsible for a single piece of the program | Good division of roles.  These classes should be broken out into separate files.   | 
Classes are loosely coupled | Yes, although your `ReservationSystem` class has too much functionality.  I would shift some of the duties, like seeing if two date ranges overlap into other classes.    | 
**Wave 1** |
List rooms |  Check in the `get_all_rooms` method | 
Reserve a room for a given date range | Check in the `reservation` method.  | 
List reservations for a given date |  I don't see a method to get reservations for a given date.  |
Calculate reservation price |  No price information in the classes. | 
Invalid date range produces an error | Check, although it seems like you are raising `ArgumentError` incorrectly.  Check comments in the code.    | 
Test coverage | ~ 90%  | 
**Wave 2** |
View available rooms for a given date range | Check in `get_open_rooms` | 
Reserving a room that is not available produces an error | Missing a test for this.  The method seems to be there.  | 
Test coverage |
**Wave 3** |
Create a block of rooms |  Check in `make_new_block`.  Good limit it to 5 rooms.  I would name the class to `BlockOfRooms` instead of `Block_Of_Rooms`  |
Check if a block has rooms | Yes, but there should be a method in Block to do so.  |  
Reserve a room from a block | Check `reserve_room_from_block`  |
Test coverage | Still 90% | 
**Additional Feedback** | There's a lot here and you obviously put a lot of work into this.  I would suggest spending a bit more time next project keeping your project organized.  Breaking classes into separate files and looking a large blocks of code and breaking them into methods.  In testing checking edge cases like room numbers outside the range and testing each of the ArgumentErrors you were trying to raise.  That would detect other bugs in your code.      |

**Review by CheezItMan — COMMENTED** [review]
I may have some more comments later.
